### ISSUE TITLE
Group PM shows up on PMs screen with blank name, just commas

### ISSUE THREAD

**Issue by gnprice** [zulip/zulip-mobile#2722]
On a user's screen when they were showing me something else in the app, I saw the PMs screen in the main nav. Out of several conversations, one of them was labeled just ", ," -- comma, space, comma -- where the names of the other people should go. It looked like what you might get with a group PM thread with 3 other people, if their names were all the empty string.

I've seen this symptom myself in the past a couple of times. It'd be good to nail down the cause and fix it.

This was on an iPhone in 13.4.88, the latest. The org was zephyr.zulipchat.com, which is somewhat special on the server side -- but I don't see an obvious way that would cause this issue (and when I've seen this myself, it wasn't on zephyr), and in any event this is a buggy thing to show.

**Comment by gnprice** [comment]
Here's the plan I'd suggest for how to do this:

First, some cleanups to help make this part of the code easier to understand and work with. This will make the later steps easier, and will also help us out every time we come back to this code in the future.
* Put an appropriate type annotation on the `conversations` prop on the various components here -- `Object[]` isn't very helpful.
* ConversationGroup isn't really a group, and it's not about just any conversation. Let's give it a better name, and at least a one-line JSDoc comment saying what it really is. A good one would be `GroupPmConversationItem` and a comment saying
```
/** A list item describing one group PM conversation. */
```
* Similarly ConversationList (`PmConversationList`) and ConversationsCard (`PmConversationsCard`).

Then, to actually fix the issue:
* I like having the users' names get looked up from the `users` state at render time -- that keeps the `users` state as the single source of truth about who a user is.
* The issue, then, is just that that state is missing some users that we have messages with.
* I don't have a full understanding of the flows that cause the `users` state to update, and why we don't have this state or what logic we might have that should cause us to have it. Clearly we don't consistently have it now. If we did fix things so that we did, there would probably still sometimes be bugs, causing gaps.
* On the other hand, as #1879 exploited, we have information about these users right there in `display_recipient` on the message. We can use that to fill these gaps.
* The key thing I'd like to do differently from the approach in #1879 is: let's fill the gaps *in the `users` state itself*, our single source of truth.
* I take inspiration for this from the webapp, which follows this approach. See `extract_people_from_message` in `people.js`.
* Like the webapp (see the `report_late_add` call), we can also use this in the future to log when this happens, so it helps us resolve the causes of these gaps too.

**Comment by borisyankov** [comment]
This is not caused by gaps in the users state. (in very rare cases this can be true, but for this to happen a user should have just been created and for some reason the data not updated)
 
It is caused by us having the list of conversations before we have loaded the users data at boot time.
Thus, no need to update the 'users' state which is being reloaded anyways, just the 'conversations'

**Comment by gnprice** [comment]
> This is not caused by gaps in the users state. (in very rare cases this can be true, but for this to happen a user should have just been created and for some reason the data not updated)
>
> It is caused by us having the list of conversations before we have loaded the users data at boot time.

Interesting. Can you explain why you believe this is the right diagnosis?

FWIW, I have never seen this go away while I'm looking at it. When I saw it the other day, it was on the screen for several seconds; and we went to other screens and back to that one over a period of at least a minute, and the symptom didn't change.


> Thus ... just use the data for the 'group conversations' display.

I don't understand what this means. Can you elaborate? (Just use which data? For which display -- do you mean the PMs list, `ConversationList`?)

**Comment by borisyankov** [comment]
For more context about what data is available:

We only intentionally do have gaps in the `messages` state.
For any other data like 'users', 'subscriptions', 'unread' etc. we expect to have all the information at all times. 

There are two cases where we don't:
* on startup or account switching - have no data about some or all of these entries
* on disconnect or when data is cached - we have data that was fully loaded for a past time, but now is not up-to-date. In that case, we fully reload the data.

**Comment by gnprice** [comment]
Got it, thanks.

My hypothesis is there's a bug which causes us to be missing something in the `users` state. That wouldn't be very surprising.

The cases you describe sound like they'd lead to something different from this aspect of the observed behavior:

> I have never seen this go away while I'm looking at it. When I saw it the other day, it was on the screen for several seconds; and we went to other screens and back to that one over a period of at least a minute, and the symptom didn't change.

But maybe I'm not imagining those cases right. How would you expect it to look in those cases?

In any case, regardless of the cause (these two cases, or some unknown bug), we're rendering some information about these PM conversations while we don't have the relevant bits of `users` data. We can either fill in the data (using what's right there in the message data), or refuse to show the conversations, or show something broken like this. If the lack of data were resolving itself quickly (like 100ms), we could maybe block for it... but it's not. So filling in the data seems like the best plan to me.

**Comment by gnprice** [comment]
One other detail, which is touched on in my OP above but to make it more explicit: When I saw this symptom on a user's screen the other day on the PMs screen, there were several other PM conversations there (I think all of them 1:1 PMs) where the names displayed just fine.

Looking at `ConversationList.js`, it looks like we get the names for 1:1 PMs from the `users` state (via the `usersByEmail` prop there) in the same way we do for group PMs. So I think that means we had some of the `users` state we should, but not all of it.

**Comment by borisyankov** [comment]
> My hypothesis is there's a bug which causes us to be missing something in the users state. That wouldn't be very surprising.

That would be surprising to me though. Our user data handling is quite simple. While this does not rule out bugs... two things:
 * I have never encountered such a but
 * I am not sure what case would produce a lot of missing data, in the way we process it currenly (we mostly load all users or update them when a new user arrives)

**Comment by borisyankov** [comment]
My assumption is that all the user data was missing and you might have been mistaken that there were 1:1 PMs shown. If that is the case - we have no 'partial data' issue. If that is not the case, we should definitely reproduce the issue and fix with priority.

Separately... I don't think filling the data is optimal, as we don't have all the data for a user, just a bit of it, so we will need to introduce concepts like supporting partial data everywhere we have user data, and also reconciling data differences - partial knowledge vs incoming full data from the server. Nothing too complicated, my feeling is that keeping it simple is better from maintainability perspective.

In case of missing user data, we do not perform a narrow. So at least with our current implementation, hiding a group conversation is the best approach.

**Comment by gnprice** [comment]
> My assumption is that all the user data was missing and you might have been mistaken that there were 1:1 PMs shown.

No, I'm quite sure there were other conversations shown on the screen in addition to this broken one. It's right there in my original report:

> Out of several conversations, one of them was labeled just [...]

I don't know the cause, but that was the observed behavior and it doesn't do any good to deny it.

**Comment by gnprice** [comment]
One possible cause for having incomplete data on users: the bug we diagnosed as the likely cause of #2800, where we don't notice that we've lost our event queue on the server. If that happened and then another user showed up, we wouldn't learn about them.

Another possible cause: we don't seem to handle `EVENT_USER_UPDATE` actions. So if a user changes their email address, we won't learn about that until a refresh, even if we still have a live event queue.

Those are a couple of known cases where we'd have partial data and would stay in that state indefinitely. There may of course be unknown bugs that cause that too.

**Comment by gnprice** [comment]
This was fixed in 851a13724364730d567e21c3096c8713a281745e, which we merged 2018-07-24, a few days after the last comments above.

With that fix, the symptoms described in this issue should be impossible. If we do lack data about a user in a PM conversation, we'll just not show that conversation on this screen rather than show something bogus.

This possible cause of not having such data still exists:

> Another possible cause: we don't seem to handle `EVENT_USER_UPDATE` actions. So if a user changes their email address, we won't learn about that until a refresh, even if we still have a live event queue.

That's now tracked as part of #3408 .
